src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -259,83 +259,121 @@ private JsonPropertyInfo GetPropertyWithUniqueAttribute(Type attributeType, Dict
 
         public JsonPropertyInfo GetProperty(ReadOnlySpan<byte> propertyName, ref ReadStackFrame frame)
         {
-            JsonPropertyInfo info = null;
-
             // Keep a local copy of the cache in case it changes by another thread.
             PropertyRef[] localPropertyRefsSorted = _propertyRefsSorted;
 
+            ulong key = GetKey(propertyName);
+
             // If there is an existing cache, then use it.
             if (localPropertyRefsSorted != null)
             {
-                ulong key = GetKey(propertyName);
-
                 // Start with the current property index, and then go forwards\backwards.
                 int propertyIndex = frame.PropertyIndex;
 
                 int count = localPropertyRefsSorted.Length;
                 int iForward = Math.Min(propertyIndex, count);
                 int iBackward = iForward - 1;
 
-                while (iForward < count || iBackward >= 0)
+                for (;;)
                 {
                     if (iForward < count)
                     {
-                        if (TryIsPropertyRefEqual(localPropertyRefsSorted[iForward], propertyName, key, ref info))
+                        PropertyRef propertyRef = localPropertyRefsSorted[iForward];
+                        if (key == propertyRef.Key)
                         {
-                            return info;
+                            if (propertyName.Length <= PropertyNameKeyLength ||
+                                // We compare the whole name, although we could skip the first 7 bytes (but it's not any faster)
+                                propertyName.SequenceEqual(propertyRef.Info.Name))
+                            {
+                                return propertyRef.Info;
+                            }
                         }
                         ++iForward;
-                    }
 
-                    if (iBackward >= 0)
+                        if (iBackward >= 0)
+                        {
+                            propertyRef = localPropertyRefsSorted[iBackward];
+                            // This path is the same code as above; copied here instead of creating a method since this is a hot path.
```

Comment:
Why not have a method that is marked as aggressively inlined, instead?

Reply:
The IL was different (slower). I'll re-verify and share the results.